lib/zlib.js

```diff
@@ -466,164 +449,134 @@ function emitCloseNT(self) {
-      self._offset = 0;
-      self._buffer = Buffer.allocUnsafe(self._chunkSize);
-    }
+module.exports = exports = {
+  Deflate: createFn(Deflate),
+  Inflate: createFn(Inflate),
+  Gzip: createFn(Gzip),
+  Gunzip: createFn(Gunzip),
+  DeflateRaw: createFn(DeflateRaw),
+  InflateRaw: createFn(InflateRaw),
+  Unzip: createFn(Unzip),
+
+  // Convenience methods.
+  // compress/decompress a string or buffer in one step.
+  deflate: createConvenienceMethod(Deflate, false),
+  deflateSync: createConvenienceMethod(Deflate, true),
+  gzip: createConvenienceMethod(Gzip, false),
+  gzipSync: createConvenienceMethod(Gzip, true),
+  deflateRaw: createConvenienceMethod(DeflateRaw, false),
+  deflateRawSync: createConvenienceMethod(DeflateRaw, true),
+  unzip: createConvenienceMethod(Unzip, false),
+  unzipSync: createConvenienceMethod(Unzip, true),
+  inflate: createConvenienceMethod(Inflate, false),
+  inflateSync: createConvenienceMethod(Inflate, true),
+  gunzip: createConvenienceMethod(Gunzip, false),
+  gunzipSync: createConvenienceMethod(Gunzip, true),
+  inflateRaw: createConvenienceMethod(InflateRaw, false),
+  inflateRawSync: createConvenienceMethod(InflateRaw, true)
+};
 
-    if (availOutAfter === 0) {
-      // Not actually done.  Need to reprocess.
-      // Also, update the availInBefore to the availInAfter value,
-      // so that if we have to hit it a third (fourth, etc.) time,
-      // it'll have the correct byte counts.
-      inOff += (availInBefore - availInAfter);
-      availInBefore = availInAfter;
+exports.createDeflate = exports.Deflate;
+exports.createInflate = exports.Inflate;
+exports.createDeflateRaw = exports.DeflateRaw;
+exports.createInflateRaw = exports.InflateRaw;
+exports.createGzip = exports.Gzip;
+exports.createGunzip = exports.Gunzip;
+exports.createUnzip = exports.Unzip;
 
-      if (!async)
-        return true;
-
-      var newReq = self._handle.write(flushFlag,
-                                      chunk,
-                                      inOff,
-                                      availInBefore,
-                                      self._buffer,
-                                      self._offset,
-                                      self._chunkSize);
-      newReq.callback = callback; // this same function
-      newReq.buffer = chunk;
-      return;
-    }
-
-    if (!async)
-      return false;
+Object.defineProperty(exports, 'constants', {
+  configurable: false,
+  enumerable: true,
+  value: constants
+});
 
-    // finished with the chunk.
-    cb();
-  }
-};
+function bkeyGetter(bkey) {
+  return () => constants[bkey];
+}
+const bkeys = Object.keys(constants);
+for (var bk = 0; bk < bkeys.length; bk++) {
+  var bkey = bkeys[bk];
+  Object.defineProperty(exports, bkey, {
+    enumerable: true,
+    get: internalUtil.deprecate(bkeyGetter(bkey),
+                                `zlib.${bkey} is deprecated. Please use ` +
+                                `zlib.constants.${bkey}`, 'DEP00XX')
```

Comment:
Ok, let's get more input from @nodejs/ctc on it. I have no problem pulling this back out for now, I'd just like for us to move away from this *at some point*.